### DISCUSSION TITLE
async BackgroundTasks run in the foreground

### DISCUSSION THREAD

**Discussion by ebanner** [fastapi/fastapi#8195] [Questions] [answered]
**Describe the bug**
async `BackgroundTasks` are not running as background tasks. My response blocks until the task finishes.

**To Reproduce**
1. Create a file `run.py` with contents

```python
from fastapi import FastAPI, BackgroundTasks

import time
import asyncio

app = FastAPI()

async def sleep_for_10_seconds():
    start = time.time()
    while True:
        now = time.time()
        if now - start > 10: 
            break
        asyncio.sleep(1)
    print(time.time(), 'slept for 10 seconds')


@app.get("/")
async def read_root(background_tasks: BackgroundTasks):
    print(time.time(), 'read_root')
    background_tasks.add_task(sleep_for_10_seconds)
    return {"Hello": "World"}
```

2. Run with `uvicorn run:app --reload`
3. Run `curl http://localhost:8000`

**Expected behavior**
`{"Hello": "World"}` returns immediately. Instead it takes 10 seconds.

**Environment:**
- macOS
- `fastapi==0.27.0`
- ` uvicorn==0.8.4`
- Python 3.6.6 :: Anaconda, Inc.

**Additional context**
I originally thought this problem was confined to `asyncio.ensure_future()` in #403, but I'm getting the same behavior with `fastapi.BackgroundTasks`.

sync tasks return immediately (i.e. don't define with async).

**Comment by euri10** [comment]
try with `def sleep_for_10_seconds():` instead of `async def sleep_for_10_seconds():`

**Comment by dmontagu** [comment]
@euri10 That may be a reasonable short term workaround, but I don’t think you should have to pay the threadpool overhead just to get background tasks to finish after returning the response.. If this is the behavior on the most recent version of FastAPI I think it merits a fix.

@ebanner I’ll look into this as soon as I can (hopefully tomorrow), but in the mean time could you confirm that it isn’t fixed by updating to the most recent version of fastapi?

**Comment by euri10** [comment]
/agree

you can also use it the Starlette way but you have to declare your JSONResponse and won't benefit from the docs goodies

```

import time
import asyncio

app = FastAPI()

async def sleep_for_10_seconds():
    start = time.time()
    while True:
        now = time.time()
        if now - start > 10:
            break
        asyncio.sleep(1)
    print(time.time(), 'slept for 10 seconds')


@app.get("/")
async def read_root():
    print(time.time(), 'read_root')
    background_tasks = BackgroundTasks()
    background_tasks.add_task(sleep_for_10_seconds)
    return JSONResponse({"Hello": "World"}, background=background_tasks)
```

**Comment by dmontagu** [comment, marked as answer]
@euri10 @ebanner I think I found the problem:

```python
async def sleep_for_10_seconds():
    start = time.time()
    while True:
        now = time.time()
        if now - start > 10:
            break
        await asyncio.sleep(1)  # <-- HERE (was missing the await)
    print(time.time(), 'slept for 10 seconds')
```
After making this change, it seems to behave properly.

**Comment by euri10** [comment]
evil :+1:

**Comment by ebanner** [comment]
Ah, my mistake!

Thank you for pointing it out.

**Comment by tiangolo** [comment]
Thanks @dmontagu and @euri10 for the help!

Thanks @ebanner for reporting back and closing the issue.

**Comment by gary-x0pa** [comment]
I realised if you replaced asyncio.sleep with time.sleep it blocks the request anyone has a reason why async background tasks does not run in separate thread and how to potentially solve this problem

**Reply by Ali-Parandeh** [reply]
As soon as you mark a function with the `async` keyword, you are running it on an event loop. FastAPI won't run it inside a threadpool anymore whereas normal background task functions get scheduled to run on a threadpool instead
